src/psyclone/psyir/nodes/psy_data_node.py

```diff
@@ -133,12 +135,38 @@ def __init__(self, ast=None, children=None, parent=None, options=None):
             # FIXME: This may not be a good solution
             symtab = SymbolTable()
 
+        # Ensure that we have a container symbol for the API access
```

Comment:
Generally the node constructors do not modify other parts of the tree or the symbol tables. But I see this psy_data constructor (not just what you added but what already was here) tries to do lots of things (insert children, parts of gen_code, ...).

I am not too familiar with how this is used but from a high level view I would expect the transformation that inserts the PSyDataNode to perform these manipulations and the constructor, if anything, just check if given arguments exist/are valid. Is there any reason all these need to be inside the constructor in this case?

That said, if moving these steps to the transformation is not a trivial modification, I am fine to overlooking this for the purpose of this PR.

Reply:
I think most of what I've added was necessary in order to update the way the name of the symbol associated with the profiling region is obtained. However, whether that really needs to be done in the constructor is debatable. As the comments say, it means it can be displayed in the `str()` but, since this is a variable in generated code, I'm not sure we need this. In an ideal world, would you say that most of this symbol creation would be perfomed in the `lower_to_language_leve()` method? I can certainly see that the creation of the Schedule and association of the child nodes should be done in the Transformation. In fact, we probably need a `create()` method to perform this, as we do elsewhere.
As four different classes inherit from `PSyDataNode`, I think most of these modifications should be done in a separate issue. If @hiker agrees, I'm happy to remove the setting of the variable name from the constructor (provided that it doesn't then result in lots of other changes to this PR).

Reply:
That's rather a difficult question, since I think a lot of code here comes from various sources (e.g. region name is from @rupertford , I get only blamed because I moved the code from ProfileNode to PsyData node when using psydata for profile).

1) Why there is so much done ... as far as I can see: I think mostly that's a different philosophy - I don't like to repeat code that inserts children in the transformations, and I would prefer to have the nodes take care of that. Isn't @sergisiso adding part of the consistency checking in a recent (or current) PR? This way it is guaranteed that the node sits at the right place. While I don't mind that much (I think we agreed to disagree :) ) - I want to point out the fact that we already ran into problems because some constructors do NOT fully initialise a node: gocean loop nodes don't have bounds properly defined, meaning we can't do any bounds checking in the validation function of a transformation. IIRC it was pointed out that this was difficult to fix now (and I checked it certainly requires more than just calling the functions that define the bounds, there are other things missing preventing this from working). So, by moving parts of the initialisation out of the node, we risk (am not saying that we indeed are) that we create the same situation again.

2) I think the handling of the profiling name is originally from @rupertford, git is just blaming me because I moved Rupert's code from Profile into PsyData when I made the ProfileNode based on PSyData. But Rupert might have a better idea of why the name is set there. 

In general I prefer to have nodes to be 'ready' to be used anywhere once they are constructed.

Reply:
No worries for now, I just mentioned it partly to understand the reason it was done this way, and partly to see if the change was easier for somebody more familiar with this code, but I agree this is not part of the PR and can be discussed separately.

@hiker :

>  Isn't @sergisiso adding part of the consistency checking in a recent (or current) PR? 

The parent link consistency PR indeed removes some part of this code and from lots of transformations :)

> In general I prefer to have nodes to be 'ready' to be used anywhere once they are constructed.

> we already ran into problems because some constructors do NOT fully initialise a node

I would in general agree with you in these points, but we needed a constructor that initiases incomplete Node (w/o its connections) for the way the parsing and some PSyIR construction is done which is essentially a top-down approach. To have a more safe constructor we added the Node.create() static method with includes the children and symbol checking.

It's true that the PSyData family of nodes won't appear when the parsing from source and maybe we won't use the thin constructor of this nodes at all, but I think its still worth having the same expectations of what they do like in the other nodes.

Reply:
OK, I think we all have the same goals in mind, perhaps just different routes to get there :-) I've opened #1185 for this.